### PR TITLE
Lots of updates to the Mutect2 wdl

### LINKED ISSUE #2949: Hardcoded memory in M2 wdl
@LeeTL1220 commented on [Thu Mar 02 2017](https://github.com/broadinstitute/gatk-protected/issues/925)

These should be parameters to the entire workflow. 

Perhaps the runtime block should have an extra 1GB of RAM.

---

@takutosato commented on [Fri Mar 03 2017](https://github.com/broadinstitute/gatk-protected/issues/925#issuecomment-284081259)

I don't think this is a good idea. We'd be giving the user to many knobs to turn. 99% of users won't need to touch these parameters, and they clutter the .json and .wdl. mutect2.wdl is already so cluttered I don't wanna look at it. I think all runtime parameters should be hardcoded. 

---

@vdauwera commented on [Sat Mar 04 2017](https://github.com/broadinstitute/gatk-protected/issues/925#issuecomment-284129344)

If you organize the inputs into blocks and keep all such knobs together at the end it's not too bad. A lot of our users will need to be able to tweak those settings -- and the others can ignore them. 

See here for an example of how we do it: https://github.com/broadinstitute/wdl/blob/develop/scripts/broad_pipelines/PublicPairedSingleSampleWf_160927.inputs.json

---

@LeeTL1220 commented on [Tue Mar 14 2017](https://github.com/broadinstitute/gatk-protected/issues/925#issuecomment-286607287)

So we are a slimmer version of what @vdauwera has.

@takutosato I agree with your frustrations, but then we have to hardcode to the worst case, which will be quite expensive (in the cloud), underutilized (in all backends), and have trouble dispatching (in SGE).  

---

@LeeTL1220 commented on [Tue Mar 14 2017](https://github.com/broadinstitute/gatk-protected/issues/925#issuecomment-286607541)

@vdauwera I will happily accept comments on our json templates.  

https://github.com/broadinstitute/gatk-protected/tree/master/scripts/mutect2_wdl

---

@LeeTL1220 commented on [Tue Mar 14 2017](https://github.com/broadinstitute/gatk-protected/issues/925#issuecomment-286607739)

@davidbenjamin @takutosato The more I think about it, the more important I think this issue is.

---

@vdauwera commented on [Tue Mar 14 2017](https://github.com/broadinstitute/gatk-protected/issues/925#issuecomment-286613100)

Yeah we need to parameterize the heck out of all our WDLs. If anything, the example I linked to is not parameterized nearly as much as I'd like (it's derived from the prod pipeline so we're a bit constrained). 

It's not that much clutter if you make those parameters task-level and organize the JSONs clearly. And it makes it waaaay easier for people to adjust what they need without touching the WDL itself. This becomes even more important once you move the WDL into a platform like FireCloud, where changing the WDL is a huge pain, whereas tweaking parameters (via a method config) is trivial.

### LINKED ISSUE #2962: Clean up Mutect2.wdl
@takutosato commented on [Mon Mar 27 2017](https://github.com/broadinstitute/gatk-protected/issues/959)

Mutect2.wdl is a monolithic beast that does way more than what it's designed to do i.e. run Mutect2. It's good style to split standalone tasks (Mutect2, Oncotator, OxoG filter) into modules. On the other hand, in the cloud it's cheaper to have all tasks in one wdl. We should find a good balance between these opposing forces.

---

@davidbenjamin commented on [Sun Apr 09 2017](https://github.com/broadinstitute/gatk-protected/issues/959#issuecomment-292795106)

@takutosato I recently made a few modest (and orthogonal) steps to clean the wdl (PR #974).  Let me know if the rebase gets messy, since it will be my fault for causing the mess.

### LINKED ISSUE #2970: Allow CollectSequencingArtifactMetrics to be skipped if user specifies an input file for the Orientation Bias filter.
@LeeTL1220 commented on [Wed Apr 19 2017](https://github.com/broadinstitute/gatk-protected/issues/989)

In other words, if a user specifies the file to use, do not run the task.

### LINKED ISSUE #3270: M2 WDL has hardcoded storage values in runtime blocks that are a bit large
When running a fairly modest number of samples (40) with scatter count of 100, I can max out/nearly max out the DSDE Methods quota.

We should make a parameter (with a default) for controlling the amount of disk to use.

### LINKED ISSUE #3341: M2 WDL should take the comments from the FC WDL
comment block in the FC WDL is a good description, though not exactly the same as the GATK M2 WDL.
Also, the input parameters are re-ordered in a way that makes sense.
```
## Copyright Broad Institute, 2017
## 
## This WDL workflow runs GATK4 Mutect 2 on a single tumor-normal pair or on a single tumor sample, 
## and performs additional filtering and functional annotation tasks. 
##
## Main requirements/expectations :
## - One analysis-ready BAM file (and its index) for each sample 
##
## Description of inputs :
##
## ** Runtime ** (requires Docker; to use on-premises without Docker, change gatk4_jar from String to File)
## gatk4_jar: path to the java jar file containing GATK 4 (beta.2 or later) in the specified docker
## picard_jar: path to retrieve a Picard jar (will be replaced by a docker image in a future version)
## m2_docker, oncotator_docker: docker images to use for GATK4 Mutect2 and for Oncotator
## preemptible_attempts: how many preemptions to tolerate before switching to a non-preemptible machine (on Google)
##
## ** Workflow options **
## intervals: genomic intervals (will be used for scatter)
## scatter_count: number of parallel jobs to generate when scattering over intervals
## artifact_modes: filtering options
## m2_extra_args, m2_extra_filtering_args: additional arguments for Mutect2 calling and filtering (optional)
## is_run_orientation_bias_filter: if true, run the orientation bias filter post-processing step
## is_run_oncotator: if true, annotate the M2 VCFs using oncotator (to produce a TCGA MAF)
##
## ** Primary inputs **
## ref_fasta, ref_fasta_index, ref_dict: reference genome, index, and dictionary
## tumor_bam, tumor_bam_index, and tumor_sample_name: BAM, index and sample name for the tumor sample (sample name used for output naming)
## normal_bam, normal_bam_index, and normal_sample_name: BAM, index and sample name for the normal sample (optional if running tumor-only)
##
## ** Primary resources ** (optional but strongly recommended)
## pon, pon_index: optional panel of normals in VCF format containing probable technical artifacts (false positves)
## gnomad, gnomad_index: optional database of known germline variants (see http://gnomad.broadinstitute.org/downloads)
## variants_for_contamination, variants_for_contamination_index: VCF of common variants with allele frequencies fo calculating contamination
##
## ** Secondary resources ** (for optional tasks)
## onco_ds_tar_gz, default_config_file: Oncotator datasources and config file
## sequencing_center, sequence_source: metadata for Oncotator
##
## Outputs :
## - One VCF file and its index with primary filtering applied; secondary filtering and functional annotation if requested.
##
## Cromwell version support 
## - Successfully tested on v27
##
## LICENSING : 
## This script is released under the WDL source code license (BSD-3) (see LICENSE in 
## https://github.com/broadinstitute/wdl). Note however that the programs it calls may 
## be subject to different licenses. Users are responsible for checking that they are
## authorized to run all programs before running this script. Please see the docker 
## pages at https://hub.docker.com/r/broadinstitute/* for detailed licensing information 
## pertaining to the included programs.

workflow Mutect2 {
  # Runtime
  String gatk4_jar
  File picard_jar
  String m2_docker
  String oncotator_docker
  Int preemptible_attempts
  # Workflow options
  Int scatter_count
  File? intervals  
  Array[String] artifact_modes
  String? m2_extra_args
  String? m2_extra_filtering_args
  Boolean is_run_orientation_bias_filter
  Boolean is_run_oncotator
  # Primary inputs 
  File ref_fasta
  File ref_fasta_index
  File ref_dict
  File tumor_bam
  File tumor_bam_index
  String tumor_sample_name
  File? normal_bam
  File? normal_bam_index
  String? normal_sample_name
  # Primary resources
  File? pon
  File? pon_index
  File? gnomad
  File? gnomad_index
  File? variants_for_contamination
  File? variants_for_contamination_index
  # Secondary resources / inputs
  File? onco_ds_tar_gz
  File? default_config_file
  String? sequencing_center
  String? sequence_source
  
...snip...
```

### LINKED ISSUE #3625: Remove download of picard.jar from .travis.yml and update Mutect2 WDL
Once https://github.com/broadinstitute/gatk/pull/3620/ is in, we should be able to remove the download of picard.jar from .travis.yml, and change the M2 WDL to no longer depend having access to it. Workflow calls to picard tools can be replaced with calls to the same tools in GATK, although the argument syntax will have to change from picard style to Barclay style ("I=..." to "-I ...").

### LINKED ISSUE #4142: `mutect2_opt.wdl` should be merged with `mutect2.wdl` without NIO
We should probably have one round of merging of mutect2_opt.wdl into mutect2.wdl, without merging in the changes to enable NIO.

The changes in mutect2_opt.wdl should supersede what is in mutect2.wdl.  And make sure that NIO is not brought in (yet)

@jsotobroad

### PR DESCRIPTION
**Pull request by davidbenjamin**
Closes #2962 (clean M2 wdl)
Closes #2949 (hardcoded memory)
Closes #2970 (optional user-supplied sequencing artifact metrics)
Closes #3341 (comments from FC wdl)
Closes #3270 (smarter disk requirements)
Closes #3625 (remove picard jar)
Closes #4142 (merge improvements from mutect2_opt wdl)

@takutosato This ought to be the last significant change to the M2 wdl for a while.  I hope.

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/Mutect2FilteringEngine.java [hunk 1]
```diff
@@ -30,6 +30,8 @@ private void applyContaminationFilter(final M2FiltersArgumentCollection MTFAC, f
         final Genotype tumorGenotype = vc.getGenotype(tumorSample);
         final double[] alleleFractions = GATKProtectedVariantContextUtils.getAttributeAsDoubleArray(tumorGenotype, VCFConstants.ALLELE_FREQUENCY_KEY,
                 () -> new double[] {1.0}, 1.0);
+        final int maxIndex = MathUtils.maxElementIndex(alleleFractions);
```
Comment (takutosato) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/Mutect2FilteringEngine.java]: You probably didn't mean to include this

File: scripts/mutect2_wdl/mutect2.wdl [hunk 2]
```diff
@@ -1,28 +1,60 @@
...
+##
+## Main requirements/expectations :
+## - One analysis-ready BAM file (and its index) for each sample
+##
+## Description of inputs:
+##
+## ** Runtime ** 
+## gatk_docker, oncotator_docker: docker images to use for GATK4 Mutect2 and for Oncotator
+## preemptible_attempts: how many preemptions to tolerate before switching to a non-preemptible machine (on Google)
+## gatk_override: (optional) local file or Google bucket path to a GATK 4 java jar file to be used instead of the GATK 4 jar
```
Comment (takutosato) [inline scripts/mutect2_wdl/mutect2.wdl]: GATK 4 -> GATK4 (either is fine, just be consistent)

File: scripts/mutect2_wdl/mutect2.wdl [hunk 3]
```diff
@@ -1,28 +1,60 @@
...
+##
+## ** Primary inputs **
+## ref_fasta, ref_fai, ref_dict: reference genome, index, and dictionary
+## tumor_bam, tumor_bam_index: BAM and index for the tumor sample
+## normal_bam, normal_bam_index: BAM and index for the normal sample
+##
+## ** Primary resources ** (optional but strongly recommended)
+## pon, pon_index: optional panel of normals in VCF format containing probable technical artifacts (false positves)
+## gnomad, gnomad_index: optional database of known germline variants (see http://gnomad.broadinstitute.org/downloads)
+## variants_for_contamination, variants_for_contamination_index: VCF of common variants with allele frequencies fo calculating contamination
```
Comment (takutosato) [inline scripts/mutect2_wdl/mutect2.wdl]: allele frequencies for*

**Comment by davidbenjamin** [comment]
> It's far more important/effective to run the script itself to check that it runs, which I think you've done many a time.

That's exactly my approach to reviewing wdl PRs.  It's on the submitter the verify the results and on the reviewer to check the code's quality.

File: scripts/mutect2_wdl/mutect2.wdl (hunk 2 above)
Comment (davidbenjamin) [reply scripts/mutect2_wdl/mutect2.wdl]: done

File: scripts/mutect2_wdl/mutect2.wdl (hunk 3 above)
Comment (davidbenjamin) [reply scripts/mutect2_wdl/mutect2.wdl]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/Mutect2FilteringEngine.java (hunk 1 above)
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/Mutect2FilteringEngine.java]: whoops, slipped in from that other branch

**Comment by davidbenjamin** [comment]
@takutosato I fixed the errors you pointed out.

**Comment by takutosato** [comment]
Great, please merge when the tests pass

**Pull request merged**